Enhancement/875 Transition from WP Acceptance to Cypress E2E tests

faisal-alvi opened the pull request:
### Description of the Change

Adding Cypress E2E tests and removing WP Acceptance tests.

#### Added the following tests:

- [x] Settings test.

<!-- Enter any applicable Issues here. Example: -->
Closes #875

### Alternate Designs

n/a

### Possible Drawbacks

n/a

### Verification Process

Check if all Cypress tests are running locally + at Github actions.

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/10up/.github/blob/trunk/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests passed.

<!-- _NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._ -->

### Changelog Entry

> Added - Cypress E2E tests.

### Credits

Props @faisal-alvi

faisal-alvi committed:
enhancement/875 Transition from WP Acceptance to Cypress E2E tests

faisal-alvi committed:
adding composer step in yml file

faisal-alvi committed:
upgrading the dep in lock file to test if that fixes issue

faisal-alvi committed:
reverting back manual changes from lock file

faisal-alvi committed:
adding de/activate plugin test

faisal-alvi committed:
using the util function to visit the page

faisal-alvi committed:
using `activatePlugin` and `deactivatePlugin`

faisal-alvi committed:
removing unnecessary script

faisal-alvi committed:
a custom plugin to disable SSL check in HTTP requests

faisal-alvi committed:
a custom plugin to disable SSL check in HTTP requests

faisal-alvi committed:
loading the custom plugin

faisal-alvi committed:
set port to 80 and try visiting the network plugins page

faisal-alvi committed:
fix the function not found issue

faisal-alvi committed:
updating the port, proper way

faisal-alvi committed:
adding constants to the tests config

faisal-alvi committed:
adding Basic Auth plugin

faisal-alvi committed:
adding basic auth plugin files

faisal-alvi committed:
upgrading `@wordpress/env` dep

faisal-alvi committed:
add new tests + remove custom plugin

faisal-alvi committed:
removing plugin from the list

faisal-alvi committed:
overwrite `.htaccess` for multisite

faisal-alvi committed:
enable the cypress video

faisal-alvi committed:
activate plugins network wide

faisal-alvi committed:
overriding the `got` dep

dinhtungdu commented on .wp-env.json:
```diff
@@ -0,0 +1,18 @@
+    ".",
+    "./tests/Basic-Auth-master"
+  ],
+  "config": {
+    "WP_ALLOW_MULTISITE": true
+  },
```
We need to remove this line for the `multisite-convert` command to work. Otherwise, we will get this error:

```
> wp-env run tests-cli "wp core multisite-convert --title='Distributor Multisite'"

ℹ Starting 'wp core multisite-convert --title='Distributor Multisite'' on the tests-cli container. 

[+] Running 2/0
 ⠿ Container c6d05ada51b8eda479c4ae247b278e53-tests-mysql-1      Running                                   0.0s
 ⠿ Container c6d05ada51b8eda479c4ae247b278e53-tests-wordpress-1  Running                                   0.0s
The network already exists.
✔ Ran `wp core multisite-convert --title='Distributor Multisite'` in 'tests-cli'. (in 1s 42ms)
```

faisal-alvi committed:
remove the `WP_ALLOW_MULTISITE` from config

faisal-alvi committed:
try to fix isse: Error [ERR_REQUIRE_ESM]: require() of ES Module

faisal-alvi committed:
revert the last change

faisal-alvi committed:
check versions + update the package lock file

faisal-alvi committed:
trying to fix https://github.com/10up/distributor/runs/7362812514?check_suite_focus=true

faisal-alvi committed:
reverting last change

faisal-alvi committed:
upgrading the node version

faisal-alvi committed:
upgrading node for linting

faisal-alvi committed:
adding a step to remove package lock file
to fix https://github.com/10up/distributor/runs/7363099128?check_suite_focus=true

faisal-alvi committed:
fix issue of invalid `-`

faisal-alvi committed:
reverting prev changes

peterwilsoncc commented:
I'll move this to the 2.0.0 milestone as 1.7.0 likely to be released early next week.

faisal-alvi committed:
removing override for `got`
to fix:
npm ERR! git@github.com: Permission denied (publickey).
npm ERR! fatal: Could not read from remote repository.

Sidsector9 committed:
regenrate package-lock.json

peterwilsoncc commented:
We discussed this issue in the 10up APAC call today.

It appears that Cypress is not building either because:
* different users have different versions of node/npm
* Cypress requires a specific version range of node/npm

To ensure users are using the same version of node, it would be good to add an `.nvmrc` file to this plugin and document that development requires https://github.com/nvm-sh/nvm -- there are [docs in Gutenberg](https://github.com/WordPress/gutenberg/blob/trunk/docs/getting-started/devenv/README.md#node-development-tools) that may be useful inspiration.

If Cypress requires a specific version of node, then let's aim for 16, 14 or 12 in order of preference -- each of them being LTS versions.

jeffpaul commented:
I'm wildly opinionated that we should do what is necessary to ensure we can utilize Cypress for e2e tests on Distributor since we're spending time on building up our Cypress utilities across other 10up OSS projects and would be sub-optimal to have a different e2e testing framework just with Distributor.  So if that means adding an `.nvmrc` file here, then so be it please proceed (also check to see if that should somehow be called out in the release instructions or if its a one-time addition).

faisal-alvi committed:
Merge branch 'develop' into enhancement/875

faisal-alvi committed:
adding `.npmrc` file

faisal-alvi committed:
fixing issue: every step must define a `uses` or `run` key

faisal-alvi committed:
adding `v` before version number

faisal-alvi committed:
trying to fix : You have an error in your yaml syntax on line 26

faisal-alvi committed:
reverting the node version back to `8.9.10`

faisal-alvi committed:
removing the recent steps to see if all workflows start again

faisal-alvi committed:
trying cehcking downgrade cypress to see if all workflows start again

iamdharmesh committed:
Merge branch 'develop' of https://github.com/10up/distributor into enhancement/875

iamdharmesh committed:
update node version and few action versions in lint.yml

iamdharmesh committed:
run "npm audit fix"

iamdharmesh committed:
upgrade @wp/env to 5.0.0

dkotter commented:
As discussed today, the failing Dependency Review check is already solved in #916, so that doesn't need addressed here.

In addition, it was decided that we should proceed with merging this and any additional changes need (like additional test coverage, removing WP Acceptance tests, etc) can be handled in follow up PRs.
